Sources/PackageSyntax/ManifestRewriter.swift

```diff
@@ -0,0 +1,822 @@
+            diagnosticsEngine.emit(.missingProduct(name: product))
+            throw Diagnostics.fatalError
+        }
+
+        let productTargetsFinder = ArrayExprArgumentFinder(expectedLabel: "targets")
+        productTargetsFinder.walk(productNode)
+
+        guard case .found(let productTargets) = productTargetsFinder.result else {
+            diagnosticsEngine.emit(.missingArgument(name: "targets", parent: "product '\(product)'"))
+            throw Diagnostics.fatalError
+        }
+
+        let newManifest = productTargets.withAdditionalElementExpr(ExprSyntax(
+            StringLiteralExprSyntax(target)
+        )).root
+
+        self.editedSource = newManifest.as(SourceFileSyntax.self)!
+    }
+
+    private func findOrCreateTargetsList(in packageInitExpr: FunctionCallExprSyntax) throws -> ArrayExprSyntax {
+        let targetsFinder = ArrayExprArgumentFinder(expectedLabel: "targets")
+        targetsFinder.walk(packageInitExpr.argumentList)
+
+        let targetsNode: ArrayExprSyntax
+        switch targetsFinder.result {
+        case .found(let existingTargets):
+            targetsNode = existingTargets
+        case .missing:
+            // We didn't find a targets section, so insert one.
+            let argListWithTargets = EmptyArrayArgumentWriter(argumentLabel: "targets",
+                                                              followingArgumentLabels:
+                                                              "swiftLanguageVersions",
+                                                              "cLanguageStandard",
+                                                              "cxxLanguageStandard")
+                .visit(packageInitExpr.argumentList)
+
+            // Find the inserted section.
+            let targetsFinder = ArrayExprArgumentFinder(expectedLabel: "targets")
+            targetsFinder.walk(argListWithTargets)
+            guard case .found(let newTargets) = targetsFinder.result else {
+                fatalError("Could not find just-inserted targets array")
+            }
+            targetsNode = newTargets
+        case .incompatibleExpr:
+            diagnosticsEngine.emit(.incompatibleArgument(name: "targets"))
+            throw Diagnostics.fatalError
+        }
+
+        return targetsNode
+    }
+
+    private func findPackageInit() throws -> FunctionCallExprSyntax {
+        // Find Package initializer.
+        let packageFinder = PackageInitFinder()
+        packageFinder.walk(editedSource)
+        switch packageFinder.result {
+        case .found(let initFnExpr):
+            return initFnExpr
+        case .foundMultiple:
+            diagnosticsEngine.emit(.multiplePackageInits)
+            throw Diagnostics.fatalError
+        case .missing:
+            diagnosticsEngine.emit(.missingPackageInit)
+            throw Diagnostics.fatalError
+        }
+    }
+}
+
+// MARK: - Syntax Visitors
+
+/// Package init finder.
+final class PackageInitFinder: SyntaxVisitor {
```

Comment:
must we use reference types and inheritance here? what does SyntaxVisitor provide that requires it?